### PR TITLE
[asan] add address sanitizer support for syncd

### PR DESCRIPTION
**Pull request by Yakiv-Huryk**
* add --enable-asan for syncd compilation
* add SIGTERM handler that calls __lsan_do_leak_check() to generate a
report

Signed-off-by: Yakiv Huryk <yhuryk@nvidia.com>

### CODE REVIEW

File: syncd/main.cpp [hunk 1]
```diff
@@ -1,10 +1,32 @@
...
+#endif
+
 int syncd_main(int argc, char **argv);
 
 int main(int argc, char **argv)
 {
     SWSS_LOG_ENTER();
 
+#if defined(ASAN_ENABLED)
+    if (signal(SIGTERM, sigterm_handler) == SIG_ERR)
```
[the remark this answers is not in the document]
Comment (Yakiv-Huryk) [reply syncd/main.cpp:24]: On the Mellanox platform, we don't stop the syncd via syncd_request_shutdown (https://github.com/Azure/sonic-buildimage/blob/master/files/scripts/syncd.sh#L55).
So, in order for us to generate the asan report, we need to catch the SIGTERM and generate it explicitly via __lsan_do_leak_check().
Another benefit of this approach is that in such case, the report will not contain any records of memory with the program's lifetime.

**Comment by liat-grozovik** [comment]
@kcudnik , @prsunny could you please signoff?

**Comment by kcudnik** [comment]
would it be possible to make this as separate file and signal handler marka as __attribute__((constructor)) ? And include this file for compilation only when asan is enabled?

**Review by kcudnik — DISMISSED** [review]

**Review by oleksandrivantsiv — APPROVED** [review]

**Comment by Yakiv-Huryk** [comment]
> would it be possible to make this as separate file and signal handler marka as **attribute**((constructor)) ? And include this file for compilation only when asan is enabled?

Good idea, thanks. 
I've updated the PR - https://github.com/Azure/sonic-sairedis/pull/1020/commits/7a0ad215a05f4522ecbe087ea2c1532170352dc7

**Review by kcudnik — APPROVED** [review]

**Pull request merged**